File: app/src/main/java/com/mainstreetcode/teammate/viewmodel/MappedViewModel.java
package com.mainstreetcode.teammate.viewmodel;


import android.annotation.SuppressLint;

import com.mainstreetcode.teammate.util.FunctionalDiff;
import com.mainstreetcode.teammate.model.Message;
import com.mainstreetcode.teammate.model.Model;
import com.mainstreetcode.teammate.notifications.Notifier;
import com.mainstreetcode.teammate.util.ErrorHandler;
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import androidx.arch.core.util.Function;
import androidx.core.util.Pair;
import androidx.recyclerview.widget.DiffUtil;
import io.reactivex.Flowable;
import io.reactivex.Single;

import static com.mainstreetcode.teammate.model.Message.fromThrowable;
import static com.tunjid.androidbootstrap.functions.collections.Lists.findLast;

public abstract class MappedViewModel<K, V extends Differentiable> extends BaseViewModel {

    private AtomicInteger pullToRefreshCount = new AtomicInteger(0);
    private Notifier.NotifierFactory factory = new Notifier.NotifierFactory();

    MappedViewModel() {}

    abstract Class<V> valueClass();

    public abstract List<Differentiable> getModelList(K key);

    abstract Flowable<List<V>> fetch(K key, boolean fetchLatest);

    Flowable<Differentiable> checkForInvalidObject(Flowable<? extends Differentiable> source, K key, V value) {
        return source.cast(Differentiable.class).doOnError(throwable -> checkForInvalidObject(throwable, value, key));
    }

    Single<Differentiable> checkForInvalidObject(Single<? extends Differentiable> source, K key, V value) {
        return source.cast(Differentiable.class).doOnError(throwable -> checkForInvalidObject(throwable, value, key));
    }

    public Flowable<DiffUtil.DiffResult> getMany(K key, boolean fetchLatest) {
        return fetchLatest ? getLatest(key) : getMore(key);
    }

    public Flowable<DiffUtil.DiffResult> getMore(K key) {
        return FunctionalDiff.of(fetch(key, false).map(this::toDifferentiable), getModelList(key), this::preserveList)
                .doOnError(throwable -> checkForInvalidKey(throwable, key));
    }

    public Flowable<DiffUtil.DiffResult> refresh(K key) {
        return FunctionalDiff.of(fetch(key, true).map(this::toDifferentiable), getModelList(key), this::pullToRefresh)
                .doOnError(throwable -> checkForInvalidKey(throwable, key))
                .doOnTerminate(() -> pullToRefreshCount.set(0));
    }

    private Flowable<DiffUtil.DiffResult> getLatest(K key) {
        return FunctionalDiff.of(fetch(key, true).map(this::toDifferentiable), getModelList(key), this::preserveList)
                .doOnError(throwable -> checkForInvalidKey(throwable, key));
    }

    public void clearNotifications(V value) {
        clearNotification(notificationCancelMap(value));
    }

    @SuppressLint("CheckResult")
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public void clearNotifications(K key) {
        Flowable.fromIterable(getModelList(key))
                .map(this::notificationCancelMap)
                .subscribe(this::clearNotification, ErrorHandler.EMPTY);
    }

    private List<Differentiable> pullToRefresh(List<Differentiable> source, List<Differentiable> additions) {
        if (pullToRefreshCount.getAndIncrement() == 0) source.clear();

        preserveList(source, additions);
        afterPullToRefreshDiff(source);
        return source;
    }

    void afterPullToRefreshDiff(List<Differentiable> source) {}

    void onInvalidKey(K key) {}

    void onErrorMessage(Message message, K key, Differentiable invalid) {
        if (message.isInvalidObject()) getModelList(key).remove(invalid);
    }

    Pair<Model, Class> notificationCancelMap(Differentiable identifiable) {
        if (!(identifiable instanceof Model)) return new Pair<>(null, null);
        Model model = (Model) identifiable;
        return new Pair<>(model, model.getClass());
    }

    @SuppressWarnings("unchecked")
    private void clearNotification(Pair<Model, Class> pair) {
        if (pair.first == null || pair.second == null) return;
        Notifier notifier = factory.forClass(pair.second);
        if (notifier != null) notifier.clearNotifications(pair.first);
    }

    void checkForInvalidObject(Throwable throwable, V model, K key) {
        Message message = fromThrowable(throwable);
        if (message != null) onErrorMessage(message, key, model);
    }

    Date getQueryDate(boolean fetchLatest, K key, Function<V, Date> dateFunction) {
        if (fetchLatest) return null;

        V value = findLast(getModelList(key), valueClass());
        return value == null ? null : dateFunction.apply(value);
    }

    private void checkForInvalidKey(Throwable throwable, K key) {
        Message message = fromThrowable(throwable);
        boolean isInvalidModel = message != null && !message.isValidModel();

        if (isInvalidModel) onInvalidKey(key);
    }

    final List<Differentiable> toDifferentiable(List<V> source) {
        return new ArrayList<>(source);
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/viewmodel/MappedViewModel.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.viewmodel


import androidx.recyclerview.widget.DiffUtil
import com.mainstreetcode.teammate.model.Chat
import com.mainstreetcode.teammate.model.Event
import com.mainstreetcode.teammate.model.Game
import com.mainstreetcode.teammate.model.JoinRequest
import com.mainstreetcode.teammate.model.Media
import com.mainstreetcode.teammate.model.Message
import com.mainstreetcode.teammate.model.Model
import com.mainstreetcode.teammate.model.Role
import com.mainstreetcode.teammate.model.Team
import com.mainstreetcode.teammate.model.Tournament
import com.mainstreetcode.teammate.model.User
import com.mainstreetcode.teammate.model.toMessage
import com.mainstreetcode.teammate.notifications.NotifierProvider
import com.mainstreetcode.teammate.util.FunctionalDiff
import com.mainstreetcode.teammate.util.asDifferentiables
import com.tunjid.androidx.recyclerview.diff.Differentiable
import io.reactivex.Flowable
import io.reactivex.Single
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

abstract class MappedViewModel<K, V : Differentiable> internal constructor() : BaseViewModel() {

    private val pullToRefreshCount = AtomicInteger(0)

    internal abstract fun valueClass(): Class<V>

    abstract fun getModelList(key: K): MutableList<Differentiable>

    internal abstract fun fetch(key: K, fetchLatest: Boolean): Flowable<List<V>>

    internal fun checkForInvalidObject(source: Flowable<out Differentiable>, key: K, value: V): Flowable<Differentiable> =
            source.cast(Differentiable::class.java).doOnError { throwable -> checkForInvalidObject(throwable, value, key) }

    internal fun checkForInvalidObject(source: Single<out Differentiable>, key: K, value: V): Single<Differentiable> =
            source.cast(Differentiable::class.java).doOnError { throwable -> checkForInvalidObject(throwable, value, key) }

    fun getMany(key: K, fetchLatest: Boolean): Flowable<DiffUtil.DiffResult> =
            if (fetchLatest) getLatest(key) else getMore(key)

    fun getMore(key: K): Flowable<DiffUtil.DiffResult> = FunctionalDiff.of(
            fetch(key, false).map(::asDifferentiables),
            getModelList(key),
            this::preserveList
    )
            .doOnError { throwable -> checkForInvalidKey(throwable, key) }

    fun refresh(key: K): Flowable<DiffUtil.DiffResult> = FunctionalDiff.of(
            fetch(key, true).map(::asDifferentiables),
            getModelList(key),
            this::pullToRefresh
    )
            .doOnError { throwable -> checkForInvalidKey(throwable, key) }
            .doOnTerminate { pullToRefreshCount.set(0) }

    fun swap(from: K, to: K): Flowable<DiffUtil.DiffResult> = FunctionalDiff.of(
            Flowable.fromCallable { getModelList(from) }.map(::asDifferentiables),
            getModelList(to)
    ) { _, copiedFrom -> copiedFrom }
            .concatWith(Flowable.defer { refresh(to) })

    private fun getLatest(key: K): Flowable<DiffUtil.DiffResult> = FunctionalDiff.of(
            fetch(key, true).map(::asDifferentiables),
            getModelList(key),
            this::preserveList
    )
            .doOnError { throwable -> checkForInvalidKey(throwable, key) }

    fun clearNotifications(value: V) = clearNotification(itemToModel(value))

    fun clearNotifications(key: K) =
            getModelList(key).mapNotNull(this::itemToModel).forEach(this::clearNotification)

    private fun pullToRefresh(source: List<Differentiable>, additions: List<Differentiable>): List<Differentiable> {
        val next = if (pullToRefreshCount.getAndIncrement() == 0) listOf() else source

        return afterPullToRefreshDiff(preserveList(next, additions))
    }

    internal open fun afterPullToRefreshDiff(source: List<Differentiable>): List<Differentiable> = source

    internal open fun onInvalidKey(key: K) {}

    internal open fun onErrorMessage(message: Message, key: K, invalid: Differentiable) {
        if (message.isInvalidObject) getModelList(key).remove(invalid)
    }

    internal open fun itemToModel(identifiable: Differentiable): Model<*>? =
            if (identifiable is Model<*>) identifiable
            else null

    private fun clearNotification(model: Model<*>?) = when (model) {
        is User -> NotifierProvider.forModel(User::class.java).clearNotifications(model)
        is Team -> NotifierProvider.forModel(Team::class.java).clearNotifications(model)
        is Role -> NotifierProvider.forModel(Role::class.java).clearNotifications(model)
        is Chat -> NotifierProvider.forModel(Chat::class.java).clearNotifications(model)
        is Game -> NotifierProvider.forModel(Game::class.java).clearNotifications(model)
        is Media -> NotifierProvider.forModel(Media::class.java).clearNotifications(model)
        is Event -> NotifierProvider.forModel(Event::class.java).clearNotifications(model)
        is Tournament -> NotifierProvider.forModel(Tournament::class.java).clearNotifications(model)
        is JoinRequest -> NotifierProvider.forModel(JoinRequest::class.java).clearNotifications(model)
        else -> Unit
    }

    internal fun checkForInvalidObject(throwable: Throwable, model: V, key: K) {
        val message = throwable.toMessage()
        if (message != null) onErrorMessage(message, key, model)
    }

    internal open fun getQueryDate(fetchLatest: Boolean, key: K, dateFunction: (V) -> Date): Date? {
        if (fetchLatest) return null

        val value = getModelList(key).filterIsInstance(valueClass()).lastOrNull()
        return if (value == null) null else dateFunction.invoke(value)
    }

    private fun checkForInvalidKey(throwable: Throwable, key: K) {
        val message = throwable.toMessage()
        val isInvalidModel = message != null && !message.isValidModel

        if (isInvalidModel) onInvalidKey(key)
    }
}